source/common/router/config_impl.cc

```diff
@@ -383,6 +383,41 @@ RouteConstSharedPtr PathRouteEntryImpl::matches(const Http::HeaderMap& headers,
   return nullptr;
 }
 
+RegexRouteEntryImpl::RegexRouteEntryImpl(const VirtualHostImpl& vhost,
+                                         const envoy::api::v2::Route& route,
+                                         Runtime::Loader& loader)
+    : RouteEntryImplBase(vhost, route, loader),
+      regex_(std::regex{route.match().regex(), std::regex::optimize}) {}
+
+void RegexRouteEntryImpl::finalizeRequestHeaders(Http::HeaderMap& headers) const {
+  RouteEntryImplBase::finalizeRequestHeaders(headers);
+
+  const Http::HeaderString& path = headers.Path()->value();
+  const char* query_string_start = strchr(path.c_str(), '?');
+  if (query_string_start == nullptr) {
+    query_string_start = path.c_str() + path.size();
+  }
+  ASSERT(std::regex_search(path.c_str(), query_string_start, regex_));
+  std::string matched_path(path.c_str(), query_string_start);
+  finalizePathHeader(headers, matched_path);
+}
+
+RouteConstSharedPtr RegexRouteEntryImpl::matches(const Http::HeaderMap& headers,
+                                                 uint64_t random_value) const {
+  if (RouteEntryImplBase::matchRoute(headers, random_value)) {
+    const Http::HeaderString& path = headers.Path()->value();
+    const char* query_string_start = strchr(path.c_str(), '?');
+    if (query_string_start == nullptr) {
+      query_string_start = path.c_str() + path.size();
+    }
+
+    if (std::regex_search(path.c_str(), query_string_start, regex_)) {
```

Comment:
What do you think of `regex_match` vs. `regex_search`? We should probably document the behavior one way or the other.

Reply:
Yes. Also, the headers do regex match I think.

Reply:
So `regex_match` tries to match the **entire** target character sequence, while `regex_search` will match any **subsequence** of the target character sequence. I would lean towards `regex_search` because it provides a superset of the functionality of `regex_match`. (You could  make a regular expression behave like `regex_match` by using `^` and `$`). Thoughts?

Reply:
You can get the same behavior with `regex_match` by adding `.*` at beginning/end, so I don't think it matters either way, as long as we document. The question really is what is going to be the common case use of this..

Reply:
Oh right, good point. I'm not sure what the common use case would be..

Reply:
I would do whatever the existing code does, which I think is regex_match in both header matching as well as virtual clusters (please check). If you could clearly document the semantics in all three cases (with potentially a couple of examples in the docs) that would be great as there has been some confusion about how this works.

Reply:
Yep, existing code uses regex_match. I will make those changes/additions :)